### PR TITLE
Initial commit - Added bash version of mac2ip calculator

### PR DESCRIPTION
**Pull request by jasonkhanlar**

### CODE REVIEW

**Comment by colindean** [comment]
Can you change the commit summary to be something more indicative of the content of the commit? E.g. "Adds bash version of calculator" or something like that.

File: pm-commotion-mac2ip [hunk 1]
```diff
@@ -0,0 +1,127 @@
+#!/bin/bash
+#
+# Script written for PittMesh by Jason Khanlar
```
Comment (colindean) [inline pm-commotion-mac2ip]: If you want to claim copyright on this, put in a header something more like this:

```
© 2016 Jason Khanlar. All rights reserved.
Licensed under the terms of the MIT license.
```

PittMesh isn't an entity for the purposes of licensing. If you want to assign the copyright to @metamesh, that's fine: swap in `Meta Mesh Wireless Communities` for your name and add an authors block that looks like this:

```
AUTHORS
* Jason Khanlar
```

File: pm-commotion-mac2ip [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+#      ((ip2-=128))
+#    elif [ $ip2 -ge 128 ]; then
+#      ((ip2-=64))
+#    fi
+#    ip2=$((ip2 % 64 + 63))
+
+    # Pad with 0
+    if [ ${#mac4} -eq 1 ]; then mac4=0$mac4; fi
+
+#    for ip3 in `seq 255`;do
```
Comment (colindean) [inline pm-commotion-mac2ip]: While it might be useful to have the history of things here, please do remove unnecessary code that's been commented out. You can just do it in another commit on top instead of amending this commit.

File: pm-commotion-mac2ip [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+  exit
+elif [ $args -eq 6 ]; then
+  mac1=`echo $1|tr [a-z] [A-Z]`
+  mac2=`echo $2|tr [a-z] [A-Z]`
+  mac3=`echo $3|tr [a-z] [A-Z]`
+  mac4=`echo $4|tr [a-z] [A-Z]`
+  mac5=`echo $5|tr [a-z] [A-Z]`
+  mac6=`echo $6|tr [a-z] [A-Z]`
+else
+  echo "Usage: $0 <MAC address>"
```
Comment (colindean) [inline pm-commotion-mac2ip]: Extract this into a `usage()` function.

File: pm-commotion-mac2ip [hunk 4]
```diff
@@ -0,0 +1,127 @@
+#!/bin/bash
+#
+# Script written for PittMesh by Jason Khanlar
+#
+
+# Check for --listall, otherwise proceed
+
+if [ "$*" = "--listall" ]; then
```
Comment (colindean) [inline pm-commotion-mac2ip]: I think I'd like to see the body of this `if` be a separate function, or maybe even a separate script.

File: pm-commotion-mac2ip [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+  echo "  $0 dc 9f db ce 13 57"
+
+  exit
+fi
+
+# Ensure that we are working with the correct large MAC address block
+# DC-9F-DB
+
+if [ ! $mac1 = "DC" -o ! $mac2 = "9F" -o ! $mac3 = "DB" ]; then
+  echo "Invalid MAC address. Make sure it begins with DC:9F:DB"
```
Comment (colindean) [inline pm-commotion-mac2ip]: This error is a little confusing. Maybe say "Unsupported MAC address. Only Ubiquiti-assigned MACs beginning with DC:9F:DB are supported." If the tool would otherwise work, and generate valid IP addresses, maybe instead say "Unsupported MAC address. Only Ubiquiti-assigned MACs beginning with DC:9F:DB are supported. The generated IP address may not be valid." and continue to run.

File: pm-commotion-mac2ip [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+  exit
+fi
+
+# Convert last three hexadecimal octets to decimal values
+ip1=100
+ip2=`echo "ibase=16; $mac4" | bc`
+ip3=`echo "ibase=16; $mac5" | bc`
+ip4=`echo "ibase=16; $mac6" | bc`
+
+# nmap -sL -n 100.64.0.0/10
```
Comment (colindean) [inline pm-commotion-mac2ip]: This might actually be useful in a test suite for the `--list-all` function...

**Comment by jasonkhanlar** [comment]
_thump_

File: mm-commotion-mac2ip [hunk 7]
```diff
@@ -0,0 +1,105 @@
+#!/bin/bash
+#
+# � 2016 Meta Mesh Wireless Communications. All rights reserved.
```
Comment (colindean) [inline mm-commotion-mac2ip]: s/Communications/Communities

File: mm-commotion-mac2ip [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+  usage
+  exit
+fi
+
+# Ensure that we are working with the correct large MAC address block
+# DC-9F-DB
+
+if [ ! $mac1 = "DC" -o ! $mac2 = "9F" -o ! $mac3 = "DB" ]; then
+  echo "Unsupported MAC address. Only Ubiquiti-assigned MAC addresses beginning with DC:9F:DB are supported."
+  exit
```
Comment (colindean) [inline mm-commotion-mac2ip]: `exit 1`

So that it's an error. Likewise in usage.

**Comment by colindean** [comment]
Almost there!

**Comment by jasonkhanlar** [comment]
_s_h_a_k_e_

**Comment by colindean** [comment]
Great work! Next (optional) step? Write a simple test script that checks the known good output of a few inputs.

**Comment by jasonkhanlar** [comment]
What kind of inputs? What determines known good outputs?

**Comment by colindean** [comment]
If you know that the MAC address `de:ad:be:ef:00:01` will always produce an IP of `100.64.87.1`, then that's a test case†. Pick another three or four, figure them out by hand. Call the script and check its output for what you'd expect.

``` shell
script_to_test=./mm-commotion-mac2ip
failure=0

test() {
  local mac=$1
  local ip=$2

  if [[ "$($script_to_test $mac)" != "$ip" ]]; then
    echo "$mac did not resolve to $ip"
    failure=1
  fi
}

test "de:ad:be:ef:00:01" "100.64.87.1"
test "d3-4d-B3-3F-10:11" "100.64.87.101"

if [[ $failure -gt 0 ]]; then 
  echo "There were failures."
  exit 1
fi
}
```

This eventually leads to a sort of generative test, where you can generate applicable MAC addresses at random and check the output against a known working implementation. This last bit is probably more work that is worth it for now, though.

_† The MACs and IPs mentioned in this post are certainly not correctly paired._
